### PR TITLE
60 add vi helper

### PR DESCRIPTION
**Pull request by Tennessee-Wallaceh**
Initial solution to #60.

Key points:
- Creates a new `flowjax.train` submodule, for organisational purposes. The old train utils are moved to `flowjax.train.data_fit`
- New VI helper is in `flowjax.train.variational_fit`
- Doesn't include a `sample_and_log_prob` method, which can come in separate work to keep the PRs simpler

- I am currently adding some tests

### CODE REVIEW

**Comment by danielward27** [comment]
Nice, I can add in the sample_and_log_prob method.

File: flowjax/train/variational_fit.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import optax
+import equinox as eqx
+from tqdm import tqdm
+
+from flowjax.utils import Array
+from flowjax.distributions import Distribution
+
+# Variational targets accept an array of proposed samples and return an array of (usually)
+# the unormalised log posterior probability of the samples.
+VariationalTarget = Callable[[Array], Array]
```
Comment (danielward27) [inline flowjax/train/variational_fit.py:13]: I think maybe using the explicit type is better, otherwise `VariationalTarget` may show in the the documentation, and someone may think this is a class called `VariationalTarget`, rather than a type alias

File: flowjax/train/variational_fit.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+# Variational losses are functions that take a distribution, a target 
+# callable and a random key, and returns a scalar loss.
+VariationalLoss = Callable[
```
Comment (danielward27) [inline flowjax/train/variational_fit.py:17]: Same as above

File: flowjax/train/variational_fit.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    approx_density = dist.log_prob(samples).reshape(-1)
+    target_density = target(samples).reshape(-1)
+    losses = approx_density - target_density
+    return losses.mean()
+
+def variational_fit(
+    key: random.KeyArray,
+    dist: Distribution,
+    target: VariationalTarget,
+    loss_fcn: VariationalLoss = elbo_loss,
```
Comment (danielward27) [inline flowjax/train/variational_fit.py:34]: In other places in the code I use `fn` to abbreviate function, so would be nice to keep it consistent here

File: flowjax/train/variational_fit.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    """
+    @eqx.filter_jit
+    def step(dist, target, key, optimizer, opt_state):
+        loss_val, grads = eqx.filter_value_and_grad(loss_fcn)(dist, target, key)
+        updates, opt_state = optimizer.update(grads, opt_state)
+        dist = eqx.apply_updates(dist, updates)
+        return dist, opt_state, loss_val
+
+    # Set up a default optimizer if None is provided
+    if optimizer is None:
```
Comment (danielward27) [inline flowjax/train/variational_fit.py:61]: I think that being able to pass an optimizer is reasonable, although it does lack consistency with `train_flow`. Especially with learning rate being such an important parameter, it seems wrong to have a hard-coded default, that you can only override by passing an optimizer.

My initial thoughts are we have three options:
1. Mirror the `train_flow` arguments, but add another optional argument for the optimizer. If the opimizer is passed, then `learning_rate` and `clip_norm` are ignored (and document this behaviour). We could modify `train_flow` similarly to allow passing of an optimizer.
2.  Set learning_rate=None, clip_norm=None and optimizer=None in the signature. Explicitly check that if optimizer is passed, both learning_rate and clip_norm are None. Then if an optimizer isn't passed, create one with learning_rate and clip_norm (setting them to default values if None).
3. Change `train_flow` and `variational_fit` to require an optimizer, with no default, and force the user to provide an optimizer.

I think I would lean towards option 2, but let me know what you think or if you have any other ideas

File: flowjax/train/variational_fit.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+def variational_fit(
+    key: random.KeyArray,
+    dist: Distribution,
+    target: VariationalTarget,
+    loss_fcn: VariationalLoss = elbo_loss,
+    num_epochs: int = 100,
+    optimizer: Optional[optax.GradientTransformation] = None,
+    show_progress: bool = True,
+    recorder: Optional[Callable] = None,
```
Comment (danielward27) [inline flowjax/train/variational_fit.py:38]: This isn't documented. It seems to store the distribution at each step? I think I wouldn't bother having that unless there is a good reason for it.

File: flowjax/train/variational_fit.py (hunk 4 above)
Comment (danielward27) [reply flowjax/train/variational_fit.py:61]: One advantage of 3, is that we could also return the optimizer from the training scripts, and reuse them in further training, but it would make things a bit more verbose.

**Review by danielward27 — COMMENTED** [review]
I've added some comments, thanks again for adding this :)

**Comment by danielward27** [comment]
I've added the `sample_and_log_prob` method in https://github.com/danielward27/flowjax/pull/65, so the training script should be updated to use that.

**Comment by danielward27** [comment]
By the way, I have no strong feelings about "train" vs "fit", but feel that we should be consistent and use one or the other and not both.

File: flowjax/train/variational_fit.py (hunk 1 above)
Comment (Tennessee-Wallaceh) [reply flowjax/train/variational_fit.py:13]: My reason for using a type alias here is mainly because the type annotations become quite complex.
e.g for the loss function `Callable[[Distribution, Callable[[Array], Array], random.KeyArray], float]`.
Feels simpler to parse 
```
VariationalLoss = Callable[
    [Distribution, VariationalTarget, random.KeyArray], 
    float
]
```

I'm not super experienced with python typing, so not sure what best practice is here, one alternative I can think of is to explicitly annotate these as `typing.TypeAlias` and move them into a `flowjax.types` submodule? Hopefully that would alleviate confusion.

Not a massive deal, so if you're not keen I will strip these out.

File: flowjax/train/variational_fit.py (hunk 5 above)
Comment (Tennessee-Wallaceh) [reply flowjax/train/variational_fit.py:38]: Yeah this is something I find useful when running small experiments to track specific behaviour over iterations.
Will strip it out and consider if I want to make an issue to add it.

File: flowjax/train/variational_fit.py (hunk 4 above)
Comment (Tennessee-Wallaceh) [reply flowjax/train/variational_fit.py:60]: I think in the interest of flexibility I think it would be good to support passing custom optimizers - I can imagine this would be something people want to play with.
On the other hand, I think you are right that the importance of the learning rate suggests it should be easier to access.

Honestly, on reflection I will just mirror the `train_flow` arguments, then we can solve the thoughts around a custom optimiser when it is actually needed.

**Comment by Tennessee-Wallaceh** [comment]
> By the way, I have no strong feelings about "train" vs "fit", but feel that we should be consistent and use one or the other and not both.

How about `fit_to_data` and `fit_to_variational_target`?

File: flowjax/train/variational_fit.py (hunk 1 above)
Comment (danielward27) [reply flowjax/train/variational_fit.py:13]: I'm not super familiar with what the best practice is either. My inclination would be to shorten it by having it e.g. Callable[Distribution, Callable, random.KeyArray], and just document the argument well enough that it's clear that the second argument maps a sample to an (unnormalised) probability. I agree it's not particularly important so you can leave it if you like (it shouldn't break anything if we want to change it later anyway).

File: flowjax/train/variational_fit.py (hunk 5 above)
Comment (danielward27) [reply flowjax/train/variational_fit.py]: (obviously we can deal with that at a different point)

Comment (Tennessee-Wallaceh) [reply flowjax/train/variational_fit.py]: Yeah, I agree that would be a good solution.

File: flowjax/train/variational_fit.py (hunk 4 above)
Comment (Tennessee-Wallaceh) [reply flowjax/train/variational_fit.py]: that change is in place now
